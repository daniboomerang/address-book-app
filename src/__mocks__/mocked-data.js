export const usersMockedData = [
  {
    gender: 'male',
    name: { title: 'Mr', first: 'Simon', last: 'Petersen' },
    location: {
      street: { number: 3865, name: 'Gormsvej' },
      city: 'Tisvilde',
      state: 'Danmark',
      country: 'Denmark',
      postcode: 28816,
      coordinates: { latitude: '-27.4297', longitude: '-37.0144' },
      timezone: { offset: '+4:00', description: 'Abu Dhabi, Muscat, Baku, Tbilisi' },
    },
    email: 'simon.petersen@example.com',
    login: {
      uuid: 'dca63e13-5a76-4d7a-aea1-817919275f83',
      username: 'purplebird556',
      password: 'testibil',
      salt: 'uPCA1DOi',
      md5: 'ac3673267c4606e6fa5f754047d57a4a',
      sha1: 'c1e2a5b2e2daa5020a02b4ba54bc8881b4233b6c',
      sha256: '4797c6d5a893a92bccca7b09cb8c19381d909389bbfeab3a6a9ec07bc8503b23',
    },
    dob: { date: '1950-02-27T16:49:15.705Z', age: 70 },
    registered: { date: '2010-10-27T11:48:05.160Z', age: 10 },
    phone: '05143951',
    cell: '88592966',
    id: { name: 'CPR', value: '270250-8357' },
    picture: {
      large: 'https://randomuser.me/api/portraits/men/59.jpg',
      medium: 'https://randomuser.me/api/portraits/med/men/59.jpg',
      thumbnail: 'https://randomuser.me/api/portraits/thumb/men/59.jpg',
    },
    nat: 'DK',
  },
  {
    gender: 'female',
    name: { title: 'Mrs', first: 'Lilja', last: 'Jokela' },
    location: {
      street: { number: 7548, name: 'Pispalan Valtatie' },
      city: 'Keminmaa',
      state: 'Ostrobothnia',
      country: 'Finland',
      postcode: 43533,
      coordinates: { latitude: '64.2460', longitude: '136.8692' },
      timezone: { offset: '-3:00', description: 'Brazil, Buenos Aires, Georgetown' },
    },
    email: 'lilja.jokela@example.com',
    login: {
      uuid: '8e3fabe6-8b6b-4736-9c94-75b994f8d414',
      username: 'crazyleopard483',
      password: 'autumn',
      salt: 'UKKs7HYq',
      md5: '2b09bc4c28288d00a2b18dac7f76de44',
      sha1: 'de5cff136b4edfeb353ed3950f8276089cc9d79b',
      sha256: 'a673bbd39e3adc7ccec88ce2d19ed24b0cc53d1e1e852c15dcfbd37ce5d8c367',
    },
    dob: { date: '1986-09-27T06:06:51.382Z', age: 34 },
    registered: { date: '2016-07-20T20:43:33.792Z', age: 4 },
    phone: '04-844-011',
    cell: '045-891-39-73',
    id: { name: 'HETU', value: 'NaNNA006undefined' },
    picture: {
      large: 'https://randomuser.me/api/portraits/women/38.jpg',
      medium: 'https://randomuser.me/api/portraits/med/women/38.jpg',
      thumbnail: 'https://randomuser.me/api/portraits/thumb/women/38.jpg',
    },
    nat: 'FI',
  },
  {
    gender: 'female',
    name: { title: 'Miss', first: 'Elizabeth', last: 'Fitzsimmons' },
    location: {
      street: { number: 8029, name: 'Galway Road' },
      city: 'Carrigtwohill',
      state: 'Monaghan',
      country: 'Ireland',
      postcode: 43719,
      coordinates: { latitude: '3.9870', longitude: '-26.4752' },
      timezone: { offset: '+4:30', description: 'Kabul' },
    },
    email: 'elizabeth.fitzsimmons@example.com',
    login: {
      uuid: 'cd65d771-329c-4f25-ac50-1cad545ac23b',
      username: 'whitebird629',
      password: 'genius',
      salt: 'LZlhnrZf',
      md5: 'ef83c24de3ed16772a5f8f9b2129361f',
      sha1: 'bbc1d6fff8914a7db416e287826f010919f34fa4',
      sha256: '271756af4be4c0d534ab2a4d3e2b1f5208063e5e067eaa94b0eaca02fd4e2581',
    },
    dob: { date: '1965-08-23T18:45:56.510Z', age: 55 },
    registered: { date: '2009-02-01T18:49:44.676Z', age: 11 },
    phone: '051-825-5109',
    cell: '081-894-8456',
    id: { name: 'PPS', value: '5705495T' },
    picture: {
      large: 'https://randomuser.me/api/portraits/women/92.jpg',
      medium: 'https://randomuser.me/api/portraits/med/women/92.jpg',
      thumbnail: 'https://randomuser.me/api/portraits/thumb/women/92.jpg',
    },
    nat: 'IE',
  },
  {
    gender: 'female',
    name: { title: 'Ms', first: 'Grace', last: 'Hayes' },
    location: {
      street: { number: 8886, name: 'Railroad St' },
      city: 'Honolulu',
      state: 'Virginia',
      country: 'United States',
      postcode: 70240,
      coordinates: { latitude: '-22.5618', longitude: '36.5461' },
      timezone: { offset: '+8:00', description: 'Beijing, Perth, Singapore, Hong Kong' },
    },
    email: 'grace.hayes@example.com',
    login: {
      uuid: 'd3134d16-e196-46fa-a498-21eef7d78a9c',
      username: 'ticklishfrog920',
      password: 'paulie',
      salt: 'FqaoRps6',
      md5: 'b51539121dab80fa3e2ada0cc3167c9d',
      sha1: '00b9218d0778b2eb07470691830698a5d318a747',
      sha256: 'fc8a46a0156e03e570fc3389e76c2dce3c7d6768b470be7ef4715d8467713522',
    },
    dob: { date: '1971-06-03T09:42:12.926Z', age: 49 },
    registered: { date: '2002-09-17T15:44:05.886Z', age: 18 },
    phone: '(841)-092-9021',
    cell: '(824)-718-6927',
    id: { name: 'SSN', value: '542-75-2315' },
    picture: {
      large: 'https://randomuser.me/api/portraits/women/33.jpg',
      medium: 'https://randomuser.me/api/portraits/med/women/33.jpg',
      thumbnail: 'https://randomuser.me/api/portraits/thumb/women/33.jpg',
    },
    nat: 'US',
  },
  {
    gender: 'female',
    name: { title: 'Ms', first: 'Dioclécia', last: 'Campos' },
    location: {
      street: { number: 2003, name: 'Rua Pernambuco ' },
      city: 'Maracanaú',
      state: 'Rondônia',
      country: 'Brazil',
      postcode: 34040,
      coordinates: { latitude: '-77.8314', longitude: '10.5344' },
      timezone: { offset: '+8:00', description: 'Beijing, Perth, Singapore, Hong Kong' },
    },
    email: 'dioclecia.campos@example.com',
    login: {
      uuid: '7f8fc56b-dea6-4ea8-803b-44ead5e705d6',
      username: 'organicpanda259',
      password: 'spread',
      salt: '8Gx3wYi5',
      md5: '99cd9a47f8379db73808f23de608fc7f',
      sha1: 'b7432a33c95357c03a4aeae311346d1e2ef7da35',
      sha256: 'b95c97745c194505db96b236d3931ae2d0275ffdb8982144411c10ec2470db47',
    },
    dob: { date: '1973-02-08T23:44:08.300Z', age: 47 },
    registered: { date: '2008-09-10T23:20:06.420Z', age: 12 },
    phone: '(47) 2714-2607',
    cell: '(12) 8149-3353',
    id: { name: '', value: null },
    picture: {
      large: 'https://randomuser.me/api/portraits/women/82.jpg',
      medium: 'https://randomuser.me/api/portraits/med/women/82.jpg',
      thumbnail: 'https://randomuser.me/api/portraits/thumb/women/82.jpg',
    },
    nat: 'BR',
  },
  {
    gender: 'female',
    name: { title: 'Miss', first: 'ملینا', last: 'رضایی' },
    location: {
      street: { number: 986, name: 'میدان فلسطین' },
      city: 'بوشهر',
      state: 'آذربایجان غربی',
      country: 'Iran',
      postcode: 80968,
      coordinates: { latitude: '-63.7699', longitude: '-151.3513' },
      timezone: { offset: '-3:30', description: 'Newfoundland' },
    },
    email: 'mlyn.rdyy@example.com',
    login: {
      uuid: 'e3f042d3-6b3c-49f3-b963-c845381e31bd',
      username: 'angryfrog283',
      password: 'sluggo',
      salt: 'qdbuA3WR',
      md5: 'ba4c2e3baf53034f3ae83d6232bcdfe1',
      sha1: '649c083f0e63627ff63629e51945cb3a0d69785d',
      sha256: 'c0cfa9ce2940dcbfdc56857095473a966768a5a81d3e33bf2ab06b8655ed90df',
    },
    dob: { date: '1955-02-04T21:44:15.732Z', age: 65 },
    registered: { date: '2015-01-08T21:12:58.291Z', age: 5 },
    phone: '011-61653559',
    cell: '0964-916-8118',
    id: { name: '', value: null },
    picture: {
      large: 'https://randomuser.me/api/portraits/women/11.jpg',
      medium: 'https://randomuser.me/api/portraits/med/women/11.jpg',
      thumbnail: 'https://randomuser.me/api/portraits/thumb/women/11.jpg',
    },
    nat: 'IR',
  },
  {
    gender: 'male',
    name: { title: 'Mr', first: 'Victor', last: 'Martin' },
    location: {
      street: { number: 5285, name: 'King Street' },
      city: 'Newport',
      state: 'Gloucestershire',
      country: 'United Kingdom',
      postcode: 'WC6 7AY',
      coordinates: { latitude: '-53.9545', longitude: '2.4196' },
      timezone: { offset: '-7:00', description: 'Mountain Time (US & Canada)' },
    },
    email: 'victor.martin@example.com',
    login: {
      uuid: '8c4f85e5-d2ed-49b2-85d6-b4b274a1b24e',
      username: 'bigfish956',
      password: 'james1',
      salt: 'ALrBE392',
      md5: '8a3811df3c7197002c846adf97c54602',
      sha1: 'b10c7f84287a086f896d4eb4664ab9b52647806a',
      sha256: '545aaae2958c475411c3aa40c0d84a7c02d69a0345b8f5a652dcdf2e63961af5',
    },
    dob: { date: '1991-07-20T02:55:40.967Z', age: 29 },
    registered: { date: '2006-07-20T17:49:47.026Z', age: 14 },
    phone: '015242 81119',
    cell: '0784-023-762',
    id: { name: 'NINO', value: 'WE 53 03 85 P' },
    picture: {
      large: 'https://randomuser.me/api/portraits/men/68.jpg',
      medium: 'https://randomuser.me/api/portraits/med/men/68.jpg',
      thumbnail: 'https://randomuser.me/api/portraits/thumb/men/68.jpg',
    },
    nat: 'GB',
  },
];

export const fetchMockedData = {
  data: {
    results: usersMockedData,
  },
};
