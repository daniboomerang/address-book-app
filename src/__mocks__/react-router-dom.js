import React from 'react';

const rrd = require('react-router-dom');

// Just render a fragment with its children
rrd.BrowserRouter = ({ children }) => <>{children}</>;
module.exports = rrd;
