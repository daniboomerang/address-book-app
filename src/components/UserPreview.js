import React, { useState } from 'react';
import cx from 'classnames';

/**
 * It renders a user preview card
 * @param {Object} user - First name, last name, username, email and avatar.
 */
const UserPreview = ({ user }) => {
  const [isHovered, setHovered] = useState(false);
  const handleMouseEnter = () => setHovered(true);
  const handleMouseLeave = () => setHovered(false);

  const {
    name: { first: firstName, last: lastName },
    login: { username },
    email,
    picture: { large },
  } = user;

  return (
    <div
      data-testid={`user-${email}`}
      className={cx(
        'w-full block sm:flex border-4 text-center border-light my-6 transform cursor-pointer duration-200 hover:scale-110',
        {
          'border-red-600 text-red-600 rounded-md': isHovered,
          'border-blue-500': !isHovered,
        }
      )}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
    >
      <div
        className={cx(
          'flex border-b-4 py-4 sm:border-r-4 sm:border-b-0 items-center justify-center duration-200',
          {
            'border-red-600': isHovered,
            'border-blue-500': !isHovered,
          }
        )}
      >
        <div
          className={cx(
            'flex border-4 rounded-full mx-4 md:mx-8 justify-center duration-200 transform',
            {
              'w-24 h-24 border-red-600': isHovered,
              'w-20 h-20 border-blue-500': !isHovered,
            }
          )}
        >
          <img
            data-testid={large}
            className="border-primary rounded-full transform duration-200 border-4"
            src={large}
            alt={username}
          />
        </div>
      </div>
      <div className="p-4 m-auto">
        <div className="my-1 text-lg md:text-xl">
          <span data-testid={firstName} className="mr-1">
            {firstName}
          </span>
          <span data-testid={lastName} className="ml-1">
            {lastName}
          </span>
        </div>
        <div data-testid={username} className="my-1 text-base md:text-lg">
          {username}
        </div>
        <div data-testid={email} className="my-1 text-sm mb-4">
          {email}
        </div>
      </div>
    </div>
  );
};

export default UserPreview;
