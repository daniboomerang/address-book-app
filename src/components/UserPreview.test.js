import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { usersMockedData } from '../__mocks__/mocked-data';
import UserPreview from './UserPreview';

describe('UserPreview component', () => {
  const user = usersMockedData[0];
  const {
    name: { first: firstName, last: lastName },
    login: { username },
    email,
    picture: { large },
  } = user;

  it('should render correctly', () => {
    const wrapper = mount(<UserPreview user={user} />);

    expect(wrapper.find(`div[data-testid="user-${email}"]`).length).toBe(1);
    expect(wrapper.find(`span[data-testid="${firstName}"]`).text()).toBe(firstName);
    expect(wrapper.find(`span[data-testid="${lastName}"]`).text()).toBe(lastName);
    expect(wrapper.find(`img[src="${large}"]`).length).toBe(1);
    expect(wrapper.find(`div[data-testid="${username}"]`).text()).toBe(username);
    expect(wrapper.find(`div[data-testid="${email}"]`).text()).toBe(email);

    expect(wrapper).toMatchSnapshot();
  });

  it('should transition from blue to red color on hover', () => {
    const wrapper = mount(<UserPreview user={user} />);

    // 1. Find user container
    let userContainer = wrapper.find(`div[data-testid="user-${email}"]`);
    expect(userContainer).toHaveLength(1);

    // 2. Check the border and text are blue
    expect(userContainer.find('.border-blue-500').length).toBeGreaterThan(1);
    expect(userContainer.find('.text-red-600').length).toBe(0);
    expect(userContainer.find('.border-red-600').length).toBe(0);

    // 3. Hover it
    act(() => {
      wrapper.update();
      userContainer.prop('onMouseEnter')({
        currentTarget: {},
      });
    });

    act(() => {
      wrapper.update();
    });

    userContainer = wrapper.find(`div[data-testid="user-${email}"]`);
    expect(userContainer).toHaveLength(1);

    // 4. Check the border and text are red after hovering
    expect(userContainer.find('.border-blue-500').length).toBe(0);
    expect(userContainer.find('.text-red-600').length).toBe(1);
    expect(userContainer.find('.border-red-600').length).toBeGreaterThan(1);

    expect(wrapper).toMatchSnapshot();
  });
});
