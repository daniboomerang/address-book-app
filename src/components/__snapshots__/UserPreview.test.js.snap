// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPreview component should render correctly 1`] = `
<UserPreview
  user={
    Object {
      "cell": "88592966",
      "dob": Object {
        "age": 70,
        "date": "1950-02-27T16:49:15.705Z",
      },
      "email": "simmon.martin@example.com",
      "gender": "male",
      "id": Object {
        "name": "CPR",
        "value": "270250-8357",
      },
      "location": Object {
        "city": "Tisvilde",
        "coordinates": Object {
          "latitude": "-27.4297",
          "longitude": "-37.0144",
        },
        "country": "Denmark",
        "postcode": 28816,
        "state": "Danmark",
        "street": Object {
          "name": "Gormsvej",
          "number": 3865,
        },
        "timezone": Object {
          "description": "Abu Dhabi, Muscat, Baku, Tbilisi",
          "offset": "+4:00",
        },
      },
      "login": Object {
        "md5": "ac3673267c4606e6fa5f754047d57a4a",
        "password": "testibil",
        "salt": "uPCA1DOi",
        "sha1": "c1e2a5b2e2daa5020a02b4ba54bc8881b4233b6c",
        "sha256": "4797c6d5a893a92bccca7b09cb8c19381d909389bbfeab3a6a9ec07bc8503b23",
        "username": "purplebird556",
        "uuid": "dca63e13-5a76-4d7a-aea1-817919275f83",
      },
      "name": Object {
        "first": "Simmon",
        "last": "Martin",
        "title": "Mr",
      },
      "nat": "DK",
      "phone": "05143951",
      "picture": Object {
        "large": "https://randomuser.me/api/portraits/men/59.jpg",
        "medium": "https://randomuser.me/api/portraits/med/men/59.jpg",
        "thumbnail": "https://randomuser.me/api/portraits/thumb/men/59.jpg",
      },
      "registered": Object {
        "age": 10,
        "date": "2010-10-27T11:48:05.160Z",
      },
    }
  }
>
  <div
    className="w-full block sm:flex border-4 text-center border-light my-6 transform cursor-pointer duration-200 hover:scale-110 outline-none border-blue-500"
    data-testid="user-simmon.martin@example.com"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="flex border-b-4 py-4 sm:border-r-4 sm:border-b-0 items-center justify-center duration-200 border-blue-500"
    >
      <div
        className="flex border-4 rounded-full mx-4 md:mx-8 justify-center duration-200 transform w-20 h-20 border-blue-500"
      >
        <img
          alt="purplebird556"
          className="border-primary rounded-full transform duration-200 border-4"
          data-testid="https://randomuser.me/api/portraits/men/59.jpg"
          src="https://randomuser.me/api/portraits/men/59.jpg"
        />
      </div>
    </div>
    <div
      className="p-4 m-auto"
    >
      <div
        className="my-1 text-lg md:text-xl"
      >
        <span
          className="mr-1"
          data-testid="Simmon"
        >
          Simmon
        </span>
        <span
          className="ml-1"
          data-testid="Martin"
        >
          Martin
        </span>
      </div>
      <div
        className="my-1 text-base md:text-lg"
        data-testid="purplebird556"
      >
        purplebird556
      </div>
      <div
        className="my-1 text-sm mb-4"
        data-testid="simmon.martin@example.com"
      >
        simmon.martin@example.com
      </div>
    </div>
  </div>
</UserPreview>
`;

exports[`UserPreview component should transition from blue to red color on hover 1`] = `
<UserPreview
  user={
    Object {
      "cell": "88592966",
      "dob": Object {
        "age": 70,
        "date": "1950-02-27T16:49:15.705Z",
      },
      "email": "simmon.martin@example.com",
      "gender": "male",
      "id": Object {
        "name": "CPR",
        "value": "270250-8357",
      },
      "location": Object {
        "city": "Tisvilde",
        "coordinates": Object {
          "latitude": "-27.4297",
          "longitude": "-37.0144",
        },
        "country": "Denmark",
        "postcode": 28816,
        "state": "Danmark",
        "street": Object {
          "name": "Gormsvej",
          "number": 3865,
        },
        "timezone": Object {
          "description": "Abu Dhabi, Muscat, Baku, Tbilisi",
          "offset": "+4:00",
        },
      },
      "login": Object {
        "md5": "ac3673267c4606e6fa5f754047d57a4a",
        "password": "testibil",
        "salt": "uPCA1DOi",
        "sha1": "c1e2a5b2e2daa5020a02b4ba54bc8881b4233b6c",
        "sha256": "4797c6d5a893a92bccca7b09cb8c19381d909389bbfeab3a6a9ec07bc8503b23",
        "username": "purplebird556",
        "uuid": "dca63e13-5a76-4d7a-aea1-817919275f83",
      },
      "name": Object {
        "first": "Simmon",
        "last": "Martin",
        "title": "Mr",
      },
      "nat": "DK",
      "phone": "05143951",
      "picture": Object {
        "large": "https://randomuser.me/api/portraits/men/59.jpg",
        "medium": "https://randomuser.me/api/portraits/med/men/59.jpg",
        "thumbnail": "https://randomuser.me/api/portraits/thumb/men/59.jpg",
      },
      "registered": Object {
        "age": 10,
        "date": "2010-10-27T11:48:05.160Z",
      },
    }
  }
>
  <div
    className="w-full block sm:flex border-4 text-center border-light my-6 transform cursor-pointer duration-200 hover:scale-110 outline-none border-red-600 text-red-600 rounded-md"
    data-testid="user-simmon.martin@example.com"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="flex border-b-4 py-4 sm:border-r-4 sm:border-b-0 items-center justify-center duration-200 border-red-600"
    >
      <div
        className="flex border-4 rounded-full mx-4 md:mx-8 justify-center duration-200 transform w-24 h-24 border-red-600"
      >
        <img
          alt="purplebird556"
          className="border-primary rounded-full transform duration-200 border-4"
          data-testid="https://randomuser.me/api/portraits/men/59.jpg"
          src="https://randomuser.me/api/portraits/men/59.jpg"
        />
      </div>
    </div>
    <div
      className="p-4 m-auto"
    >
      <div
        className="my-1 text-lg md:text-xl"
      >
        <span
          className="mr-1"
          data-testid="Simmon"
        >
          Simmon
        </span>
        <span
          className="ml-1"
          data-testid="Martin"
        >
          Martin
        </span>
      </div>
      <div
        className="my-1 text-base md:text-lg"
        data-testid="purplebird556"
      >
        purplebird556
      </div>
      <div
        className="my-1 text-sm mb-4"
        data-testid="simmon.martin@example.com"
      >
        simmon.martin@example.com
      </div>
    </div>
  </div>
</UserPreview>
`;
